pkg/cache/listers.go

```diff
@@ -27,6 +30,60 @@ import (
 	"k8s.io/client-go/tools/cache"
 )
 
+// ListAllByCluster used to list items belongs to a cluster from Indexer.
+func ListAllByCluster(indexer cache.Indexer, cluster logicalcluster.Name, selector labels.Selector, appendFn cache.AppendFunc) error {
+	return listAllByIndexWithBackup(indexer, ClusterIndexName, ClusterIndexKey(cluster), ClusterIndexFunc, selector, appendFn)
+}
+
+// ListAllByClusterAndNamespace used to list items belongs to a cluster and namespace from Indexer.
+func ListAllByClusterAndNamespace(indexer cache.Indexer, cluster logicalcluster.Name, namespace string, selector labels.Selector, appendFn cache.AppendFunc) error {
+	if namespace == metav1.NamespaceAll {
+		return ListAllByCluster(indexer, cluster, selector, appendFn)
+	}
+	return listAllByIndexWithBackup(indexer, ClusterAndNamespaceIndexName, ClusterAndNamespaceIndexKey(cluster, namespace), ClusterAndNamespaceIndexFunc, selector, appendFn)
+}
+
+// listAllByIndexWithBackup used to list items from the Indexer using an index, or falling back to a func if no index is registered.
+func listAllByIndexWithBackup(indexer cache.Indexer, indexName, indexKey string, indexFunc cache.IndexFunc, selector labels.Selector, appendFn cache.AppendFunc) error {
+	var items []interface{}
+	var err error
+	items, err = indexer.ByIndex(indexName, indexKey)
+	if err != nil {
+		// Ignore error; do slow search without index.
+		klog.Warningf("can not retrieve list of objects using index : %v", err)
+		for _, item := range indexer.List() {
+			keys, err := indexFunc(item)
+			if err != nil {
+				return err
+			}
+			if sets.NewString(keys...).Has(indexKey) {
+				items = append(items, item)
+			}
+		}
+	}
+	return appendMatchingObjects(items, selector, appendFn)
+}
+
+func appendMatchingObjects(items []interface{}, selector labels.Selector, appendFn cache.AppendFunc) error {
+	selectAll := selector == nil || selector.Empty()
+	for _, item := range items {
+		if selectAll {
+			// Avoid computing labels of the objects to speed up common flows
+			// of listing all objects.
+			appendFn(item)
+		}
+		metadata, err := meta.Accessor(item)
+		if err != nil {
+			return err
+		}
+		if selector.Matches(labels.Set(metadata.GetLabels())) {
```

Comment:
[NPE here according to CI](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/kcp-dev_apimachinery/64/pull-ci-kcp-dev-apimachinery-main-test/1587866746075746304)